### PR TITLE
fix(security)!: replace passport-openidconnect with openid-client

### LINKED ISSUE #206: Replace `passport-openidconnect` with `openid-client`
https://github.com/GluuFederation/gluu-passport/issues/204

### PR DESCRIPTION
**Pull request by kdhttps**
BREAKING CHANGE
close #206

### CODE REVIEW

File: test/file-utils.test.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+describe('fileUtils.writeDataToFile', () => {
+  it('should be exists', () => {
+    assert.exists(fileUtils.writeDataToFile)
+  })
+
+  it('should be function', () => {
+    assert.isFunction(fileUtils.writeDataToFile)
+  })
+
+  it('writeJWKS should write content in file', async () => {
```
Comment (christian-hawk) [inline test/file-utils.test.js:32]: @kdhttps  This one is more in the integration side, as it uses persistence (file).
Maybe here we can check if the method is called correctly etc.

File: test/openid-client-helper.test.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+    it('should exist', () => {
+      assert.exists(generateJWKS)
+    })
+
+    it('should be function', () => {
+      assert.isFunction(generateJWKS, 'generateJWKS is not a function')
+    })
+
+    it('should generate jwks for provider in jwks folder', async () => {
```
Comment (christian-hawk) [inline test/openid-client-helper.test.js:29]: This one should also go to `spec.js` as we are not testing the unit.

File: test/openid-client-helper.test.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+const InitMock = require('./testdata/init-mock')
+const config = require('config')
+
+const assert = chai.assert
+const passportConfigAuthorizedResponse = config.get('passportConfigAuthorizedResponse')
+
+describe('Test OpenID Client Helper', () => {
+  const testProvider = passportConfigAuthorizedResponse.providers.find(p => p.id === 'oidccedev6privatejwt')
+  let kid = null
+  const jwksFilePath = `../server/jwks/${testProvider.id}.json`
```
Comment (christian-hawk) [inline test/openid-client-helper.test.js:16]: The basic is:
- IF this file is accessed, should go to `spec.js`
- IF this file is not accessed, should stay on `test.js`

**Review by christian-hawk — COMMENTED** [review]
Looks like a nice job mate! did some comments for your evaluation

File: test/file-utils.test.js (hunk 1 above)
Comment (kdhttps) [reply test/file-utils.test.js]: Agreed Mate

File: test/openid-client-helper.test.js (hunk 3 above)
Comment (kdhttps) [reply test/openid-client-helper.test.js]: Understood, thanks mate for info

File: test/openid-client-helper.test.js (hunk 2 above)
Comment (kdhttps) [reply test/openid-client-helper.test.js]: agree

**Comment by christian-hawk** [comment]
run tests

**Comment by kdhttps** [comment]
run tests

**Comment by kdhttps** [comment]
run tests

**Comment by kdhttps** [comment]
run tests

**Comment by kdhttps** [comment]
run tests

**Comment by kdhttps** [comment]
run tests

**Comment by kdhttps** [comment]
run tests

**Comment by christian-hawk** [comment]
idp initiated failed. run tests

**Comment by kdhttps** [comment]
@christian-hawk mate, I think it is failing at starting ` PASSPORT-SAML-DEFAULT`

```
09:38:34 ============================================================================
09:38:34 
09:38:34 STARTING: PASSPORT-SAML-DEFAULT
09:38:34 
09:38:34 ============================================================================
09:38:34 
09:38:34 - Provider type: IDP
09:38:34 - Type: passport-saml
09:38:34 - Flow: default
09:38:34 - Passport Provider ID: saml-default
09:38:34 TODO: Creating Trust Relation on t3.xx.xx automatically
```

I am not able to run test in my local. I am getting `sslerror`, during droplet creating.
```
HTTPSConnectionPool(host='t1.techno24x7.com', port=443): Max retries exceeded with url: / (Caused by SSLError(SSLError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:852)'),))
Server still not ready...
```

I am thinking to add some more time sleep so that it can reach to login page and get test pass.

and as you suggest to jenkins is disabled so not sure to run t_e_s_t on jenkins or not. Please suggest how to test?

**Comment by christian-hawk** [comment]
run tests

**Comment by christian-hawk** [comment]
we need this changes in oxTrust https://github.com/GluuFederation/oxTrust/issues/2052

**Comment by christian-hawk** [comment]
Linking dependant PRs:
- Oxtrust: https://github.com/GluuFederation/oxTrust/issues/2052
- Docs https://github.com/GluuFederation/oxTrust/issues/2052

**Comment by kdhttps** [comment]
@christian-hawk mate, I've manually tested `private_key_jwt`, `pkce` and `client_secret_post` flow with latest 4.3.0 oxtrust war. and everything is working. can we merge this PR now??

Linked PR & issue https://github.com/GluuFederation/gluu-passport/pull/217#issuecomment-804197545

**Review by christian-hawk — APPROVED** [review]

**Pull request merged**